[RFC] Add support for the "special" color used for undercurls

AdnoC opened the pull request:
Tried to solve https://github.com/neovim/neovim/issues/2040.
Found that this would also resolve https://github.com/neovim/neovim/issues/3370

Adds the previously removed special color field used to specify underline/undercurl color.

Used [this branch of neovim.as](https://github.com/AdnoC/neovim.as/tree/sp_color) to make sure it worked.

&nbsp;

I'm not sure how to add functional tests for this.

AdnoC committed:
syntax: Add support for the "special" color used for undercurls

AdnoC committed:
test: Fix tests failing

AdnoC commented on src/nvim/ui_bridge.c:
```diff
@@ -305,6 +306,18 @@ static void ui_bridge_update_bg_event(void **argv)
+static void ui_bridge_update_sp_event(void **argv)
+{
+  UI *ui = UI(argv[0]);
+  if (ui->update_sp) {
```
This check is here since the tui doesn't use the special color field.

An alternative could be just adding a stub to the tui's UI struct.

bfredl replied:
you should add a stub to the tui implementation as is already done for operations that the tui doesn't support.

justinmk commented:
👍  Hopefully someone can help out with the tests.

> Used this branch of neovim.as to make sure it worked.

Just curious, why not use https://github.com/coolwanglu/neovim-e (the origin of that fork)?

AdnoC commented:
Not sure. Its just the first thing I found.

bfredl commented:
:+1: at a quick glance our screen test infrastructure should handle added attributes like this without change. You could write a a screen test that sets some highlight group to have this attribute, and then make sure this highlight is visible somewhere. Then `screen:snapshot_util()` should pick it up like any other highlight.

KillTheMule committed:
test: Add a test for the new guisp highlighting feature.

Screen.lua needed a little cosmetical adjustment to print out the color nicely.

KillTheMule commented:
I have no idea how to integrate this into the workflow properly (hints welcome), but I've written a test at https://github.com/KillTheMule/neovim/tree/test_special_hi. Note the change to screen.lua.

bfredl commented:
A bit of inconsistent forrmatting, but otherwise that looks good. What is the workflow problem? 

Please indent the highlight specifiers like so:
```
    screen:set_default_attr_ids({
      [1] = {background = hl_colors.Search},
      [2] = {reverse = true},
      [3] = {foreground = hl_colors.Message},
    })
    screen:set_default_attr_ignore({{bold=true, foreground=hl_colors.NonText}})
```

KillTheMule commented:
Ah yeah indenting lua... is there a linter we could use or so? It's very helpful to use make lint for the C files, but I always have to a bit of guesswork with lua. I changed the indenting and force pushed again.

The workflow problem is that I would have liked to either amend this PR with my commit or send a PR to AdnoC so he could merge that and easily amend this PR. The first ist impossible afaik, but the 2nd should be doable, if I just could find the right buttons at github :)

bfredl commented:
BTW the highlight specifications in existing tests have a bit inconsistent, I try to suggest the above styles in PRs that I review (expect for `hl_colors` dict which is quite superfluous, just keep the `Screen.colors.*` ), except in the case where you only have one screen:expect where you can specify them directly to screen:expect (but with the same indention), but in general it is good to make several `screen:expect`s while one anyway have setup the relevant test conditions.
In general we are less strict with the style of tests than code. But maybe I should do a PR to clean up the style of some highlight tests, if only to make it easier to say "just copy the style of these existing tests: ..."

bfredl commented:
you can do PRs to a PR branch yes, but AdnoC could also just fetch your branch and merge/cherry-pick it locally.

AdnoC commented:
Thanks for adding tests KillTheMule. I made this PR right before Passover dinner and had to do family stuff all day.

bfredl commented on test/functional/ui/highlight_spec.lua:
```diff
@@ -194,3 +194,60 @@ describe('Default highlight groups', function()
+    screen = Screen.new(25,10)
+    screen:attach()
+    screen:set_default_attr_ignore({
+      [1] = {bold=true}, 
```
please remove the bold ignore (we only ignore the bold-blue group per policy as it would clutter every visual test with the ~-marks). Also for `attr_ignore` there is no point writing out the indices, as these are never referred.

AdnoC committed:
test: Edit the special color test

AdnoC commented:
Is this better bfredl?

justinmk commented:
>  is there a linter we could use or so? It's very helpful to use make lint for the C files, but I always have to a bit of guesswork with lua

@KillTheMule Yes, though we haven't added it to CI (travis) yet, nor do I remember any note added to the wiki. See https://github.com/neovim/neovim/pull/3687

    make testlint

> But maybe I should do a PR to clean up the style of some highlight tests, if only to make it easier to say "just copy the style of these existing tests: ..."

👍

bfredl commented on test/functional/ui/highlight_spec.lua:
```diff
@@ -194,3 +194,60 @@ describe('Default highlight groups', function()
+              {1:neovim} tabbed^    |
+      ~                        |
+      {4:-- INSERT --}             |
+      ]],{
```
this line should be indented one step less, otherwise it looks good :+1:

AdnoC replied:
Is it ok to amend my previous commit and force-push? CONTRIBUTING.md says to do that for [WIP] branches, but this is currently [RFC].

ZyX-I replied:
It is OK.

AdnoC commented:
Fixed the indentation and pushed the change.

AdnoC committed:
Linting

AdnoC commented:
Rebased to fix merge conflicts caused by https://github.com/neovim/neovim/pull/4733 and changed the appropriate commit taking into account the changes the other merge caused (Changed https://github.com/neovim/neovim/pull/4633/files#diff-cbd63c814d2878576ddab95ccb80b655R6976 to take into account rgb_sp_color).

justinmk commented:
Would like to move forward on this, can https://github.com/neovim/neovim/pull/4633#discussion_r60836052 be addressed? Stub can be empty like `tui_set_icon`.

AdnoC commented:
I didn't notice the comment. Added a stub tui_update_sp function to the tui.

justinmk commented on test/functional/ui/highlight_spec.lua:
```diff
@@ -303,3 +303,60 @@ describe('Default highlight groups', function()
+    })
+  end)
+
+  it('used for undercurls', function()
```
This test did not fail after commenting out these lines:

https://github.com/neovim/neovim/pull/4633/commits/b1273e1c082233082e4d9f7b247b241057e80a64#diff-cff0db2d9810ddbe3e259dacfece5acfR335

And this entire block: https://github.com/neovim/neovim/pull/4633/commits/b1273e1c082233082e4d9f7b247b241057e80a64#diff-cbd63c814d2878576ddab95ccb80b655R6460

What's the test actually testing?

AdnoC replied:
The test is making sure the "special" attribute of text is set according to the syntax rules + highlight groups.

https://github.com/neovim/neovim/pull/4633/commits/b1273e1c082233082e4d9f7b247b241057e80a64#diff-cff0db2d9810ddbe3e259dacfece5acfR335 wouldn't make the test fail, since update_sp is pretty much only called when resizing the screen (and when updating the 'Normal' highlight group?).

Commenting https://github.com/neovim/neovim/pull/4633/commits/b1273e1c082233082e4d9f7b247b241057e80a64#diff-cbd63c814d2878576ddab95ccb80b655R6460 out should have made the test fail though. I'll try to figure out whats going wrong and fix the test.

AdnoC replied:
The issue was that Screen wasn't checking the 'special' attributes when testing for attribute equality. I rebased off master and pushed the fixed version. Had to modify one other test to account for the change.

AdnoC committed:
test: Screen also checks 'special' when testing attribute equality

justinmk commented:
Thank you @AdnoC and @KillTheMule
